src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
```

Comment:
Yes, it is a bit different but I implemented like this because of this: https://github.com/systemd/systemd/pull/3327#discussion_r64235468. If it comes out not a good idea, it is a quick fix to going back.